lib/core/CDataFrame.cc

```diff
@@ -0,0 +1,513 @@
+};
+
+//! \brief Reserves extra columns in data frame.
+class CDataFrameRowSliceReserver {
+public:
+    CDataFrameRowSliceReserver(std::size_t numberColumns,
+                               std::size_t extraColumns)
+        : m_NumberColumns{numberColumns}, m_ExtraColumns{extraColumns} {
+    }
+
+    bool operator()(TRowSlicePtrVecCItr beginSlices, TRowSlicePtrVecCItr endSlices) {
+        for (auto i = beginSlices; i != endSlices; ++i) {
+            if ((*i)->reserve(m_NumberColumns, m_ExtraColumns) == false) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+private:
+    std::size_t m_NumberColumns;
+    std::size_t m_ExtraColumns;
+};
+
+//! Compute the default slice capacity in rows.
+std::size_t computeSliceCapacity(std::size_t numberColumns) {
+    // TODO This probably needs some careful tuning, which I haven't performed
+    // yet, and probably needs to be different for different storage strategies.
+    return std::max(1048576 / sizeof(CFloatStorage) / numberColumns, std::size_t(100));
+}
+}
+
+CDataFrame::CDataFrame(std::size_t numberColumns,
+                       std::size_t sliceCapacityInRows,
+                       EReadWriteToStorage asyncReadAndWriteToStore,
+                       const TWriteSliceToStoreFunc& writeSliceToStore)
+    : m_NumberColumns{numberColumns}, m_RowCapacity{numberColumns},
+      m_SliceCapacityInRows{sliceCapacityInRows}, m_AsyncReadAndWriteToStore{asyncReadAndWriteToStore},
+      m_WriteSliceToStore{writeSliceToStore} {
+}
+
+CDataFrame::CDataFrame(std::size_t numberColumns,
+                       EReadWriteToStorage asyncReadAndWriteToStore,
+                       const TWriteSliceToStoreFunc& writeSliceToStore)
+    : CDataFrame{numberColumns, computeSliceCapacity(numberColumns),
+                 asyncReadAndWriteToStore, writeSliceToStore} {
+}
+
+bool CDataFrame::reserve(std::size_t numberThreads, std::size_t rowCapacity) {
+    if (m_NumberColumns < rowCapacity) {
+        m_RowCapacity = rowCapacity;
+
+        std::size_t stride;
+        std::tie(numberThreads, stride) = this->numberOfThreadsAndStride(numberThreads);
+        LOG_TRACE(<< "numberThreads = " << numberThreads << " stride = " << stride);
+
+        CDataFrameRowSliceReserver reserver{
+            m_NumberColumns, m_RowCapacity - m_NumberColumns};
+
+        std::vector<std::future<bool>> reserves;
+        reserves.reserve(numberThreads);
+        std::size_t j{0};
+        for (std::size_t i = 0; i + 1 < numberThreads; ++i, j += stride) {
+            auto begin = m_Slices.begin() + j;
+            auto end = m_Slices.begin() + j + stride;
+            reserves.push_back(std::async(std::launch::async, reserver, begin, end));
+        }
+        auto begin = m_Slices.begin() + j;
+        auto end = m_Slices.end();
+        if (begin != end) {
+            reserves.push_back(std::async(std::launch::async, reserver, begin, end));
```

Comment:
Using `std::launch::async` may introduce headaches with platform differences, thus increasing the testing burden.  From https://docs.microsoft.com/en-gb/cpp/standard-library/future-functions?view=vs-2017:

> The C++ standard states that if policy is launch::async, the function creates a new thread. However the Microsoft implementation is currently non-conforming. It obtains its threads from the Windows ThreadPool, which in some cases may provide a recycled thread rather than a new one. This means that the launch::async policy is actually implemented as launch::async|launch::deferred. Another implication of the ThreadPool-based implementation is that there is no guarantee that thread-local variables will be destroyed when the thread completes. If the thread is recycled and provided to a new call to async, the old variables will still exist. We therefore recommend that you do not use thread-local variables with async.

Reply:
I agree the C++11 thread support was rushed in and doesn't feel fully thought out. Also, that is a very interesting point about clean up.

The thing I have in the back of my mind is that there are lots of features coming in the near future to the C++ standard to improve thread support. For example, once executors arrive I think std::async is going to be the right model for threading a lot of our tasks related to data frames. I don't want to do work to bring in a different library if we can avoid it.